### PR TITLE
feat(flutter-searchbox): add StareProvider widget

### PR DESCRIPTION
**Pull request by mohdashraf010897**
**PR Type** `Feature`

`Description` Add StateProvider component to `flutter_searchbox` library.

[📔 Notion](https://www.notion.so/appbase/Flutter-SearchBox-State-Provider-component-83d2df80f75d4df9a83bf583f80851e1)

https://www.loom.com/share/9bf499beebea4b8e8dc95dd97b244aec

### CODE REVIEW

File: searchbase/lib/src/types.dart [hunk 1]
```diff
@@ -125,3 +125,73 @@ class Suggestion {
         this.source != null && this.source!['_popular_suggestion'] == true;
   }
 }
+
+
+enum KeysToSubscribe {
```
Comment (bietkul) [inline searchbase/lib/src/types.dart:298]: @mohdashraf010897 You've missed the documentation for it. We should describe the usage of each key.

File: searchbase/lib/src/searchcontroller.dart [hunk 2]
```diff
@@ -922,8 +922,10 @@ class SearchController extends Base {
   /// -   `customQuery`
   ///
   subscribeToStateChanges(
-      SubscriptionFunction fn, List<String>? propertiesToSubscribe) {
-    this.stateChanges.subscribe(fn, propertiesToSubscribe);
+      SubscriptionFunction fn, List<KeysToSubscribe>? propertiesToSubscribe) {
+
+    final keysToSubscribeValues = propertiesToSubscribe?.map((item)=>item.name).toList();
```
Comment (bietkul) [inline searchbase/lib/src/searchcontroller.dart]: @mohdashraf010897 What are you doing here?
`stateChanges.subscribe` should accept the `KeysToSubscribe` type.

File: searchbase/example/basic/web/main.dart [hunk 3]
```diff
@@ -91,7 +91,7 @@ void main() {
   resultWidget.triggerDefaultQuery();
 
   resultWidget.subscribeToStateChanges((change) {
```
Comment (bietkul) [inline searchbase/example/basic/web/main.dart:93]: The `change` property could be a map of `KeysToSubscribe` and `Changes`. You would have to make this change in the definition of subscription function in searchcontroller.

File: searchbase/README.md [hunk 4]
```diff
@@ -47,7 +47,7 @@ void main() {
 
   // Build DOM when search results update
   searchController.subscribeToStateChanges((change) {
-		final results = change['results'].next;
+		final results = change[KeysToSubscribe.Results.name].next;
```
Comment (bietkul) [inline searchbase/README.md]: Check comment on `subscriber` function and update usage at all places.

File: flutter_searchbox/lib/state_provider.dart [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+      print('error $e');
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    if (widget.build != null) {
+      return widget.build!(_controllersState);
+    }
+    return SizedBox.shrink();
```
Comment (bietkul) [inline flutter_searchbox/lib/state_provider.dart:389]: Is there any reason of using `SizedBox.shrink` instead of empty container?

**Review by bietkul — CHANGES_REQUESTED** [review]
@mohdashraf010897 
1. Update change log for libraries (`searchbase`, `flutter-searchbox` and `fluttter-searchbox-ui`) to add the migration steps. We would have some breaking changes.

**Comment by mohdashraf010897** [comment]
@bietkul Kindly review the changes.
Here's the loom demo post changes
https://www.loom.com/share/28092c3e08f64ec9a89b2e558f6b2824

File: flutter_searchbox/lib/state_provider.dart [hunk 6]
```diff
@@ -0,0 +1,255 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_searchbox/flutter_searchbox.dart';
+import 'package:searchbase/searchbase.dart';
+
+class SearchControllerState {
```
Comment (bietkul) [inline flutter_searchbox/lib/state_provider.dart:6]: Documentation for class is missing.

File: flutter_searchbox/lib/state_provider.dart [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+  final Widget Function(Map<String, SearchControllerState>)? build;
+
+  StateProvider({
+    this.subscribeTo,
+    this.onChange,
+    this.build,
+    Key? key,
+  }) : super(key: key) {
+    if (build == null && onChange == null) {
+      throw ("Atleast one, build or onChange prop is required.");
```
Comment (bietkul) [inline flutter_searchbox/lib/state_provider.dart]: Use `assert`.
https://stackoverflow.com/questions/57833399/whats-the-recommended-way-in-dart-asserts-or-throw-errors

File: flutter_searchbox/lib/state_provider.dart [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+      print('error $e');
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    if (widget.build != null) {
+      return widget.build!(_controllersState);
+    }
+    return Container(
```
Comment (bietkul) [inline flutter_searchbox/lib/state_provider.dart]: Use shrink box

File: searchbase/lib/src/types.dart [hunk 9]
```diff
@@ -1,6 +1,176 @@
...
+  SortTypeChanges? SortBy;
+  ReactTypeChanges? React;
+  DefaultQueryTypeChanges? DefaultQuery;
+  DefaultQueryTypeChanges? CustomQuery;
+  BooleanTypeChanges? RequestPending;
+  RecentSearchesChanges? RecentSearches;
+  DynamicMapTypeChanges? After;
+
+  ChangesController(KeysToSubscribe key, dynamic prev, dynamic next) {
+    if (key == KeysToSubscribe.Results) {}
```
Comment (bietkul) [inline searchbase/lib/src/types.dart]: What is this condition?

**Review by bietkul — CHANGES_REQUESTED** [review]

File: flutter_searchbox/lib/state_provider.dart (hunk 7 above)
Comment (mohdashraf010897) [reply flutter_searchbox/lib/state_provider.dart]: image
is this good enough?

File: searchbase/lib/src/types.dart (hunk 9 above)
Comment (mohdashraf010897) [reply searchbase/lib/src/types.dart]: removed

File: flutter_searchbox/lib/state_provider.dart [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+    this.customQuery,
+  });
+}
+
+/// It allows you to access the current state of your widgets along with the search results.
+/// For instance, you can use this component to create results/no results or query/no query pages.
+///
+/// Examples Use(s):
+///    - perform side-effects based on the results states of various widgets.
+///    - render custom UI based on the current state of app.
```
Comment (bietkul) [inline flutter_searchbox/lib/state_provider.dart:49]: Add example.

**Review by bietkul — APPROVED** [review]

**Pull request merged**